### PR TITLE
Inclusions struct unit tests

### PR DESCRIPTION
**Pull request by Overkillus**
In follow-up to https://github.com/paritytech/polkadot-sdk/pull/1432

Some additional unit tests for the inclusion struct used in the scraper.

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3660350

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3661453

**Review by BradleyOlson64 — APPROVED** [review]
Very thorough 👍

File: polkadot/node/core/dispute-coordinator/src/scraping/tests.rs [hunk 1]
```diff
@@ -662,3 +662,251 @@ fn inclusions_per_candidate_properly_adds_and_prunes() {
...
+#[test]
+fn inclusions_initialization() {
+	let inclusions = Inclusions::new();
+
+	assert!(inclusions.inclusions_inner.is_empty(), "Expected inclusions_inner to be empty");
+	assert!(
+		inclusions.candidates_by_block_number.is_empty(),
+		"Expected candidates_by_block_number to be empty"
+	);
+}
```
Comment (tdimitrov) [inline polkadot/node/core/dispute-coordinator/src/scraping/tests.rs:677]: nit: I would have skipped this test case.

I believe your point is to have good code coverage and it makes sense in that aspect. However we know that `new` just uses `BTreeMap::new()` and there is no way to end up with a non-empty map. Imho we'd better focus on tests covering actual functionality that might be broken.

This of course is personal point of view/opinion. It's okay to ignore it!

File: polkadot/node/core/dispute-coordinator/src/scraping/tests.rs [hunk 2]
```diff
@@ -662,3 +662,251 @@ fn inclusions_per_candidate_properly_adds_and_prunes() {
...
+	inclusions.insert(candidate_hash, block_number, block_hash);
+
+	// Check inclusions_inner
+	assert!(
+		inclusions.inclusions_inner.contains_key(&candidate_hash),
+		"Expected candidate_hash to be present in inclusions_inner"
+	);
+	let inner_map = inclusions.inclusions_inner.get(&candidate_hash).unwrap();
+	assert!(
+		inner_map.contains_key(&block_number),
```
Comment (tdimitrov) [inline polkadot/node/core/dispute-coordinator/src/scraping/tests.rs:695]: nit: It's worth checking that the len of the Vec is 1 too.

File: polkadot/node/core/dispute-coordinator/src/scraping/tests.rs [hunk 3]
```diff
@@ -662,3 +662,251 @@ fn inclusions_per_candidate_properly_adds_and_prunes() {
...
+		"Expected block_number to be present for the candidate_hash in inclusions_inner"
+	);
+	let hash_set = inner_map.get(&block_number).unwrap();
```
Comment (tdimitrov) [inline polkadot/node/core/dispute-coordinator/src/scraping/tests.rs:698]: ditto: `hash_set.len() == 1`

File: polkadot/node/core/dispute-coordinator/src/scraping/tests.rs [hunk 4]
```diff
@@ -662,3 +662,251 @@ fn inclusions_per_candidate_properly_adds_and_prunes() {
...
+		hash_set.contains(&block_hash),
+		"Expected block_hash to be present for the block_number in inclusions_inner"
+	);
+
+	// Check candidates_by_block_number
+	assert!(
+		inclusions.candidates_by_block_number.contains_key(&block_number),
+		"Expected block_number to be present in candidates_by_block_number"
+	);
+	let candidate_set = inclusions.candidates_by_block_number.get(&block_number).unwrap();
```
Comment (tdimitrov) [inline polkadot/node/core/dispute-coordinator/src/scraping/tests.rs:709]: ditto for the len

File: polkadot/node/core/dispute-coordinator/src/scraping/tests.rs (hunk 1 above)
Comment (Overkillus) [reply polkadot/node/core/dispute-coordinator/src/scraping/tests.rs:677]: I see your point and it definitely wasn't necessary. It's just a habit to make an init test to see if everything indeed intializes the way we are expecting but in that case it's pretty trivial. 

Gonna leave it in as it's already written.

File: polkadot/node/core/dispute-coordinator/src/scraping/tests.rs [hunk 5]
```diff
@@ -662,3 +662,251 @@ fn inclusions_per_candidate_properly_adds_and_prunes() {
...
+	assert!(!inclusions.contains(&candidate1), "Expected candidate1 to be removed");
+	assert!(!inclusions.contains(&candidate2), "Expected candidate2 to be removed");
+	assert!(inclusions.contains(&candidate3), "Expected candidate3 to remain");
+
+	// Remove block 2 and check
+	inclusions.remove_up_to_height(&3);
+	assert!(!inclusions.contains(&candidate1), "Expected candidate1 to be removed");
+	assert!(!inclusions.contains(&candidate2), "Expected candidate2 to be removed");
+	assert!(!inclusions.contains(&candidate3), "Expected candidate3 to be removed");
+}
```
Comment (tdimitrov) [inline polkadot/node/core/dispute-coordinator/src/scraping/tests.rs:789]: Very good one! 

Another nit for additional check: Add candidate at height 3 and 4. Delete at height 5 and make sure that the candidates at height 3 and 4 are gone.

File: polkadot/node/core/dispute-coordinator/src/scraping/tests.rs (hunk 2 above)
Comment (Overkillus) [reply polkadot/node/core/dispute-coordinator/src/scraping/tests.rs:695]: Good point. Added len asserts for all the maps including all the inner ones.

**Review by tdimitrov — APPROVED** [review]
Very good work @Overkillus! A few nits from me, most of them are a matter of personal test.

`test_duplicate_insertion_same_height_different_blocks` and `inclusions_remove_with_empty_maps` are great tests to have!

File: polkadot/node/core/dispute-coordinator/src/scraping/tests.rs (hunk 1 above)
Comment (tdimitrov) [reply polkadot/node/core/dispute-coordinator/src/scraping/tests.rs:677]: Sure! Definitely doesn't hurt to have it.

File: polkadot/node/core/dispute-coordinator/src/scraping/tests.rs (hunk 5 above)
Comment (Overkillus) [reply polkadot/node/core/dispute-coordinator/src/scraping/tests.rs:816]: added

File: polkadot/node/core/dispute-coordinator/src/scraping/tests.rs [hunk 6]
```diff
@@ -662,3 +662,276 @@ fn inclusions_per_candidate_properly_adds_and_prunes() {
...
+	inclusions.remove_up_to_height(&10);
+	assert!(!inclusions.contains(&candidate1), "Expected candidate1 to be removed");
+	assert!(!inclusions.contains(&candidate2), "Expected candidate2 to be removed");
+	assert!(!inclusions.contains(&candidate3), "Expected candidate3 to be removed");
+	assert!(!inclusions.contains(&candidate4), "Expected candidate4 to be removed");
+	assert!(!inclusions.contains(&candidate5), "Expected candidate5 to be removed");
+}
+
+#[test]
+fn inclusions_duplicate_insertion_same_height_and_block() {
```
Comment (eskimor) [inline polkadot/node/core/dispute-coordinator/src/scraping/tests.rs:817]: I would be more interested if pruning works as expected for same height/same block/different block. As this was precisely what went wrong in the Kusama incident. 

Ideally we would also test whether pruning still works as desired if the same candidate was included at different heights.

**Review by eskimor — COMMENTED** [review]
A little late to the party. In general good coverage, but I would love to have pruning specifically tested for edge cases.

**Pull request merged**